java-checks/src/main/java/org/sonar/java/checks/spring/SpringBeansShouldBeAccessibleCheck.java

```diff
@@ -0,0 +1,145 @@
+  private final Set<String> scannedPackages = new HashSet<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void endOfAnalysis() {
+    AnalyzerMessageReporter reporter = (AnalyzerMessageReporter) context;
+    messagesPerClassPrefix.entrySet().stream()
+      // support sub-packages and inner classes
+      .filter(entry -> scannedPackages.stream().noneMatch(entry.getKey()::contains))
+      .forEach(entry -> entry.getValue().forEach(reporter::reportIssue));
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+
+    if (!hasSemantic() || classTree.simpleName() == null) {
+      return;
+    }
+
+    String classPackageName = extractClassPrefix(classTree.symbol().type().fullyQualifiedName());
+    SymbolMetadata classSymbolMetadata = classTree.symbol().metadata();
+
+    List<SymbolMetadata.AnnotationValue> componentScanValues = classSymbolMetadata.valuesForAnnotation(COMPONENT_SCAN_ANNOTATION);
+    if (componentScanValues != null) {
+      componentScanValues.forEach(this::addToScannedPackages);
+    } else if (hasAnnotation(classSymbolMetadata, SPRING_BOOT_APP_ANNOTATION)) {
+      scannedPackages.add(classPackageName);
+    } else if (hasAnnotation(classSymbolMetadata, SPRING_BEAN_ANNOTATIONS)) {
+      addMessageToMap(classPackageName, classTree.simpleName());
+    }
+  }
+
+  private void addMessageToMap(String classPackageName, IdentifierTree classNameTree) {
+    AnalyzerMessageReporter reporter = (AnalyzerMessageReporter) context;
+    AnalyzerMessage analyzerMessage = reporter.createAnalyzerMessage(this, classNameTree, String.format(MESSAGE_FORMAT, classNameTree));
+    messagesPerClassPrefix.computeIfAbsent(classPackageName, k -> new ArrayList<>()).add(analyzerMessage);
+  }
+
+  private void addToScannedPackages(SymbolMetadata.AnnotationValue annotationValue) {
+    if (COMPONENT_SCAN_ARGUMENTS.contains(annotationValue.name()) && annotationValue.value() instanceof ExpressionTree) {
+      ExpressionTree values = (ExpressionTree) annotationValue.value();
+      if (values.is(Tree.Kind.STRING_LITERAL)) {
+        String packageName = ConstantUtils.resolveAsStringConstant(values);
+        scannedPackages.add(packageName);
+      } else if (values.is(Tree.Kind.NEW_ARRAY)) {
+        for (ExpressionTree p : ((NewArrayTree) values).initializers()) {
+          String packageName = ConstantUtils.resolveAsStringConstant(p);
+          scannedPackages.add(packageName);
+        }
+      }
+    }
+  }
+
+  /**
+   * Returns the prefix of the class name, which:
+   * - in general, is the package name (e.g. 'foo.bar')
+   * - for inner classes, is the package name + outer class name (e.g. 'foo.bar.Outer')
+   * - for classes in the default package, is the empty string
+   */
+  private static String extractClassPrefix(String fullyQualifiedClassName) {
```

Comment:
Please use the `Convert` class for all those operations as I am really unsure your implementation would work with a class with a $ in its name.

Reply:
It does work with inner classes. I have tests for this in `../ComponentScan/A.java`.

Nevertheless, good point, I was not aware of the `Convert` class.  I added noncompliant tests as well in `../ComponentScan/C.java`.